### ISSUE TITLE
MemoryStorage - etag mismatch

### ISSUE THREAD

**Issue by stephenlautier** [dotnet/orleans#4053]
We have noticed we are getting a lot of etag mismatch issues, and we couldn't figure out why, as during dev they don't usually appear but in prod they do.

Today I had investigated a bit and finally, I found out why its happening. This is how to do it:

- Use a MemoryStorage as persistence for a stateful grain
- Set `State = null`
- `WriteStateAsync`
- Deactivate grain

On next reactivation, if you use `WriteStateAsync` it will throw an etag mismatch exception, even if you explicitly call `ReadStateAsync` and then call `WriteStateAsync` it will still throw an exception, the only way to recover it is to use `ClearStateAsync`

However if you call it when the grain is still activated, it works correctly e.g.
```cs
State = new Hero {
    Name = "Rexxar"
};
await WriteStateAsync();
State = null;
await WriteStateAsync();
State = new Hero {
    Name = "Thrall"
};
await WriteStateAsync();
```

Not entirely sure if its a bug or a misuse, i think its a bit of both, but either or ideally it can be prevented because it can lead to some quite hard to track issues.

**Comment by benjaminpetit** [comment]
Sorry for the delay, but I confirm I was able to reproduce this issue with https://github.com/benjaminpetit/orleans/commit/ee6e75d4d589853dbb7303ec1b4e0b03ef1278af

**Comment by sergeybykov** [comment]
>State = null;

Wait. State property is not supposed to be assigned by grain code, only its properties.

**Comment by sergeybykov** [comment]
I guess the public setter on `State` property allows for that. Probably no need to roll that back.

**Comment by sergeybykov** [comment]
Resolved via #4191.

**Comment by stephenlautier** [comment]
Awesome 👍
